### PR TITLE
Implement iterative average structure (Issue#2039)

### LINKED ISSUE #2039: Iterative Average structures
**Is your feature request related to a problem? Please describe.**

In a recent [paper](https://pubs.acs.org/doi/abs/10.1021/acs.jpcb.7b11988) of mine, I describe an iterative algorithm to calculate an average structure from simulations.  Below is the description. 

> As a reference structure, I used the average structure in the simulations, which for B-DNA was
obtained by RMSD aligning the inner five base pairs and averaging the simulation structures iteratively to convergence. The first frame was used for the initial alignment.

I would expect that this algorithm was already previously described. If anyone knows an earlier paper please add it to this issue so we can properly cite it. I  currently do not have time to include this properly into MDAnalysis.

*Note*: I find this a good example to show how the changes of our analysis module in the last two years allow it to quickly write up a new analysis type that is easy to reuse and share.

**Describe the solution you'd like**

This is the complete implementation I used including tests. I'm posting it here should anyone be interested in including this into the align module of MDAnalysis. If anyone wants to just use this please consider to [cite my paper](https://pubs.acs.org/doi/abs/10.1021/acs.jpcb.7b11988) when you do. The code can be used under the same license that MDAnalysis uses.

```python
class AverageStructure(base.AnalysisBase):
    """Calculate an average structure after RMSD alignment to a reference."""

    def __init__(self, mobile, ref, weights=None, calc_msf=True, **kwargs):
        """init function

        Parameters
        ----------
        mobile : mda.AtomGroup
            atomgroup that is being fitted
        ref : mda.AtomGroup
            reference structure
        weights : str, array_like (optional)
            weights that can be used. If `None` use equal weights, if `'mass'`
            use masses of ref as weights or give an array of arbitrary weights.
        calc_msf : bool (optional)
            calc mean square fluctuation
        kwargs : dict
            options for AnalysisBase to adjust iteration behavior

        """
        super(AverageStructure, self).__init__(mobile.universe.trajectory,
                                               **kwargs)
        self.mobile = mobile
        self.ref_xyz = ref.positions.copy() - ref.centroid()
        self.weights = weights
        self.calc_msf = calc_msf

    def _prepare(self):
        self.structure = np.zeros(self.mobile.positions.shape)
        self.rmsd = 0
        if self.calc_msf:
            self.msf = np.zeros(self.mobile.n_atoms)

    def _single_frame(self):
        self.mobile.translate(-self.mobile.center(self.weights))
        self.rmsd += align._fit_to(
            self.mobile.positions,
            self.ref_xyz,
            self.mobile,
            weights=self.weights,
            mobile_com=np.zeros(3),
            ref_com=np.zeros(3))[1]
        self.structure += self.mobile.positions
        if self.calc_msf:
            self.msf += np.sum(
                (self.ref_xyz - self.mobile.positions)**2, axis=1)

    def _conclude(self):
        self.structure /= self.n_frames
        self.rmsd /= self.n_frames
        if self.calc_msf:
            self.msf /= self.n_frames


def iterative_average(mobile,
                      ref,
                      weights=None,
                      niter=100,
                      eps=1e-8,
                      verbose=False,
                      **kwargs):
    """Calculate an optimal reference that is also the average structure after an
    RMSD alignment. This is done with an iterative algorithm where the previous
    average structure is used as a new reference.

    Parameters
    ----------
    mobile : mda.AtomGroup
        mobile atomgroup to find the average for
    ref : mda.AtomGroup
        initial reference structure. Positions are changed by this function!
    weights : str, array_like (optional)
        weights that can be used. If `None` use equal weights, if `'mass'` use masses of ref as weights or give an array of arbitrary weights.
    niter : int (optional)
        maximum number of iterations
    eps : float (optional)
        RMSD distance at which reference and average are assumed to be equal
    verbose : bool (optional)
        verbosity
    **kwargs : dict (optional)
        AverageStructure kwargs

    Returns
    -------
    res : AverageStructure
        AverageStructure result from the last iteration

    """
    drmsd = np.inf
    for i in range(niter):
        # found a converged structure
        if drmsd < eps:
            break

        res = AverageStructure(mobile, ref, weights, **kwargs).run()
        drmsd = mda.analysis.rms.rmsd(ref.positions, res.structure, weights=weights)
        ref.positions = res.structure

        if verbose:
            print("i = {}, rmsd-change = {:.2f}, ave-rmsd = {:.2f}, ave-msf = {:.2f}".
                  format(i, drmsd, res.rmsd, res.msf.mean()))

    return res
```

Some reference tests I have. They only check for regressions. Correctness was only validated visually so don't put too much trust into these tests.

```python
from __future__ import absolute_import

import MDAnalysis as mda
from MDAnalysisTests.datafiles import PSF, DCD

import pytest
from numpy.testing import assert_array_almost_equal

# module contains iterative average implementation
import metric


@pytest.fixture
def mobile():
    u = mda.Universe(PSF, DCD)
    u.transfer_to_memory()
    return u.select_atoms('bynum 1:10')


@pytest.fixture
def reference():
    u = mda.Universe(PSF, DCD)
    return u.select_atoms('bynum 1:10')


# Pure regression test
def test_AverageStructure(mobile, reference):
    res = metric.AverageStructure(mobile, reference, stop=10).run()
    assert_array_almost_equal(res.structure,
                              [[0.54083995, -0.42515443, -1.06207113],
                               [1.07435689, -1.16633342, -1.31354781],
                               [1.03657782, 0.31960404, -1.12080256],
                               [-0.15218723, -0.3288244, -1.698552],
                               [0.08886168, -0.68049069, 0.28910915],
                               [0.92952701, -0.77678878, 0.83458763],
                               [-0.7420783, 0.4927999, 0.85065659],
                               [-1.48584671, 0.17381404, 1.22256903],
                               [-1.06299627, 1.13438787, 0.23116275],
                               [-0.22705485, 1.25698588, 1.76688837]], decimal=5)


# Pure regression test
def test_iterative_average(mobile, reference):
    res = metric.iterative_average(mobile, reference, stop=10)
    assert_array_almost_equal(res.structure,
                              [[0.54418358, -0.42490685, -1.06077607],
                               [1.08109621, -1.18605215, -1.30346445],
                               [1.05210044, 0.32313582, -1.11259767],
                               [-0.16157775, -0.3265537, -1.70656559],
                               [0.08676427, -0.66799566, 0.28346055],
                               [0.91055919, -0.76166723, 0.8188435],
                               [-0.74196764, 0.49133729, 0.85177611],
                               [-1.48156064, 0.15684868, 1.23166217],
                               [-1.0767725, 1.13997174, 0.22508162],
                               [-0.21282517, 1.25588206, 1.77257982]], decimal=5)
````

*edit kain88-de*: Removed class unrelated to this feature from example code.

### PR DESCRIPTION
**Pull request by leonwehrhan**
Fixes #2039 

Changes made in this Pull Request:
 - Implemented average structures with iterative algorithm from DOI 10.1021/acs.jpcb.7b11988. (Issue #2039)

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [x] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
📚 Documentation preview 📚: https://mdanalysis--4524.org.readthedocs.build/en/4524/

### CODE REVIEW

**Comment by yuxuanzhuang** [comment]
Hello @leonwehrhan! Welcome to MDAnalysis! Please add #2039 directly after "Fixes" so the issue will be successfully closed after this PR is merged.

Feel free to ping us when you feel it's ready for review :)

**Comment by leonwehrhan** [comment]
Hello @yuxuanzhuang, I think the PR is ready for a review. 

Could you maybe give some guidance on how to resolve the failed automatic checks? I am not sure how to tackle the issue with the mypy linter, as my local mypy does not see any issues with the files I have touched. Also, my tests should cover every line of new code (codecov/patch 100%) so I do not know why the codecov/project test failed.

Thank you very much!

**Comment by yuxuanzhuang** [comment]
Hi, there are a few aspects up for discussion (also pinging @lilyminium, since you added the `AverageStructure`):

1) Should it be a method or an Analysis class? It also makes sense to include it as an option to run `AverageStructure` (e.g., `iterative=True`). For now, I think it's acceptable to leave it as is.

2) In scenarios where you have a large system and want to obtain the average structure of, for example, a protein, the current implementation does not offer the option to do subselections, which `AverageStructure` can.

3) The code you provided doesn't actually accomplish the task. If you check for verbosity, you can see that the RMSD is not changing at all. This is because `ref.positions = res.positions` does not actually modify the position of atoms in the reference universe. What you can do is use `ref = res.universe`, which would represent the average structure from the last round of iteration. Please double-check the output of your code.

4) The test also fails to achieve its goal (except for checking for regression). One problem is that `u1 = mda.Merge(u.select_atoms("bynum 1:10"))` turns the universe into a single-frame trajectory (check `u1.trajectory`).

Don't worry about the failing of linter and main_tests (ubuntu-latest, 3.12, true, true), they are unrelated issues.

**Comment by leonwehrhan** [comment]
Hi @yuxuanzhuang,

thank you very much for the review! I have updated the code and adressed your concerns as follows:

1. While I also think an implementation as part of the `AverageStructure` class makes sense, I have left the method as it is. I could open another issue and work on it once this PR is merged?
2. I have implemented the optional `select` keyword. As the reference structure is updated throughout the iterative calculation, the selection for the reference is set in the beginning, by transforming the reference universe into a one-frame universe of the selected atoms of the current frame of the reference. The selection for the mobile universe is passed to the `AverageStructure` class in each iteration.
3. I have implented the code you suggested. I checked the output using `verbose=True` to see the RMSD is changing. I ran `iterative_average` on `u=mda.Universe(PSF, DCD)` and visually checked that the resulting structure looked reasonable.
4. Given the implementation of the `select` keyword, it is now not necessary to use an atom-sliced universe as input for the iterative workflow, so the original universe `u` is now returned and processed in the tests. I included some regression tests and some tests checking that the output of the first 10 atoms is the same that I have visually checked.

Other things I have changed:
- The `weights` keyword could not be actually used before, because calling `rms.rmsd` would produce an error. This is fixed by calling `get_weights` and is included in the tests.
- The default `eps=1e-8` was not really useful to get to a converged structure (even after 1000 iterations), so I set it to `1e-6`.
- Updated documentation.

File: package/CHANGELOG
Comment (yuxuanzhuang) [inline package/CHANGELOG]: move it to the block for release 2.8.0. Add your name to the end of the other authors (of that release)

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]: make it optional (if not set, then use mobile as reference). change the name to `reference` to match AverageStructure.

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]: add a short code snippet as an example

Comment (yuxuanzhuang) [reply package/MDAnalysis/analysis/rms.py]: including how to retrieve the final Universe from the output

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]: res is a confusing name, I would recommend av (or avg_struc)

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]: add `versionadded` remark (see other code as examples)

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]:
```suggestion
    reference for the average structure, which is used as the reference to calculate the
```

File: package/MDAnalysis/analysis/rms.py
Comment (yuxuanzhuang) [inline package/MDAnalysis/analysis/rms.py]: add citation inside package/doc/sphinx/source/references.bib.

add :footcite:p:`xxxxx` here.

**Review by yuxuanzhuang — CHANGES_REQUESTED** [review]
Looks great! Just a few comments on the documentation side.

And please fix pep8 issue reported above.

**Comment by yuxuanzhuang** [comment]
I think this function should actually be put inside the `analysis.align` module since what it does is alignment. Could you move the code (and tests) there?

The other issue is the documentation is not showing up on the webpage (there's a preview you can find above https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/rms.html). You need to add explicitly `.. autofunction:: xxx` to the documentation at the beginning of the file.

Additionally, please leave the review unresolved so that the reviewer can determine whether the issue has indeed been addressed.

**Comment by leonwehrhan** [comment]
I have moved the function to `analysis.align` and added the remark to the documentation in the beginning of the file. The documentation of `iterative_average` now shows up on https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/align.html.

**Comment by orbeckst** [comment]
@yuxuanzhuang I added you as the person responsible for the PR.

@leonwehrhan I un-assigned you because we use the "Assignee" field to indicate who from the developer team manages review and merge.

**Review by yuxuanzhuang — APPROVED** [review]
lgtm!

File: package/CHANGELOG
Comment (orbeckst) [inline package/CHANGELOG]: This is an Enhancement. Please move into the appropriate section.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:609]: Does this render correctly? Check!

If necessary, insert blank line.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:552]: add a duecredit decorator for the paper , see eg https://github.com/MDAnalysis/mdanalysis/blob/f5335b4db7f6f746f163e3a617ae6658d30d2fb9/package/MDAnalysis/analysis/bat.py#L257-L259 (look at code in the code base for how we use [duecredit](https://docs.mdanalysis.org/stable/documentation_pages/references.html#citations-using-duecredit).

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: Follow the numpy doc string standard (see https://userguide.mdanalysis.org/stable/contributing_code.html#working-with-the-code-documentation ) and start with a single line summary:
```reST
Iteratively calculate an optimal reference that is also the average structure after an RMSD alignment.

The optimal reference is defined as average structure of a trajectory, with ...
```
and include "Iterative" there so that the function name makes more sense.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:599]: Add a sentence saying in words what the example does.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: Will this code be executed when you run the doc tests? — see #3925 

If yes, please make sure that it works.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: We never use print. Use a logger at debug level.
```python
logger.debug(f"iterative_average(): i = {i}, rmsd-change = {drmsd:.5f}, ave-rmsd = {avg_struc.results.rmsd:.5f}")
```

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:648]: Add a `logger.info()` message for the converged result that shows the final values.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [reply package/MDAnalysis/analysis/align.py]: Alternatively, use a custom progress and show this information on the progress bar.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:648]: Add an `else:` clause to the `for` loop and fail with a `RuntimeError` if the procedure did not converge:
```python
else:
   errmsg = (f"iterative_average(): Did not converge in {niter} iterations to DRMSD < {tol}. "
           f"Final average RMSD = {avg_struc.results.rmsd:.5f}")
   logger.error(errmsg)
   raise RuntimeError(errmsg)
```
You should also add a test that triggers the exception (eg by running a test with low tol and niter).

When you have an iterative algorithm it's important to at least know when it fails.

File: package/doc/sphinx/source/references.bib
Comment (orbeckst) [inline package/doc/sphinx/source/references.bib]: add DOI

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: just use  `assert_allclose` if possible, see  #3743

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: just use assert_allclose (or is there a specific reason for   assert_almost_equal ??)

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: assert_allclose

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py:597]: Implement the convergence check (see above) and add a test for it.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Looking good but we need a convergence check and we need to handle text output via logging.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: ProgressBar here would be nice!

**Comment by leonwehrhan** [comment]
@yuxuanzhuang @orbeckst I have made the requested changes.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py]: Use markup
```suggestion
    `iterative_average` can be used to obtain a :class:`MDAnalysis.Universe` with
```

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: Make this a separate test method `test_iterative_average_convergence_failure()`.

We want tests to be fine grained and testing one thing.

File: testsuite/MDAnalysisTests/analysis/test_align.py
Comment (orbeckst) [inline testsuite/MDAnalysisTests/analysis/test_align.py]: These are just running through code paths and they do not assert anything. These are only marginally useful tests.

Please split `test_iterative_average` into multiple testing methods where each method tests one thing (eg default values, different eps, weights, ...) and where you 
1. run iterative_average() _once_
2. assert that the output is as expected

This will be more code but in this way we will see exactly if one specific part of the code fails.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks for making the changes. 

My main comment is for the tests: please split your tests so that you always just test one thing as this will help tremendously in debugging if anything ever fails. Please see the comments inline.

**Comment by leonwehrhan** [comment]
@orbeckst I have changed the tests, so that every testing method only contains one call of `iterative average` and changed the documentation as you suggested.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:564]: I just checked https://mdanalysis--4524.org.readthedocs.build/en/4524/documentation_pages/analysis/align.html#MDAnalysis.analysis.align.iterative_average and the reference  `[1]` does not resolve.  Please check how to use footnote citations, e.g., as used in https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/bat.py 

I think you need to remove the reference from references and instead add a 
```reST

References
----------

.. footbibliography::

```
at the end of the module docs.

File: package/MDAnalysis/analysis/align.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/align.py:643]: add a test that runs through this block, i.e., with `verbose=True`. You don't have to check for logger output but we need to check that there are no errors in the f-string.

Comment (orbeckst) [reply package/MDAnalysis/analysis/align.py:643]: This will also help to increase coverage.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks for changing the tests. Just found two small things. Almost there!!

**Comment by leonwehrhan** [comment]
@orbeckst @yuxuanzhuang the requested changes are pushed, but it looks like auto-merge has been disabled by my commit because I do not have write access. Could one of you do the merge?

**Comment by orbeckst** [comment]
Auto-merge just means that the merge will be performed when all requirements are met
- all reviews are approving
- all required status checks passing

It's just a short cut for maintainers to essentially "queue up" the merge.

Let me have a quick look.

File: package/MDAnalysis/analysis/rms.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Why was this import added? Can we remove it again? This almost certainly creates a circular import.
```suggestion
```

File: package/MDAnalysis/analysis/rms.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/rms.py]: Why was this import added? 
```suggestion
```

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please check the imports that I flagged.

In order to remove them you can simply use my suggestions in the GH interface.

**Comment by leonwehrhan** [comment]
Thanks for the clarification! I thought I might have unintentionally halted the merge process :)

I will adress your changes now.

**Comment by leonwehrhan** [comment]
@orbeckst I have comitted the changes via your comment. The reason for these imports being there was that I originally implemented the function in `analysis.rms` instead of `analysis.align` and later moved the code (see comments above).

**Comment by orbeckst** [comment]
ok. the other test failure is not due to your stuff, see #4540

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
Thank you for your contribution! 🎉

**Comment by leonwehrhan** [comment]
Thanks a lot for guiding me through the PR!

**Comment by yuxuanzhuang** [comment]
Congrats! @leonwehrhan

**Pull request merged**
